Task-42613: Advanced filters - connected / not connected users

AzmiTouil opened the pull request:
This PR will add an advanced filter by connection status (connected or never connected) to the userManagement page.

azayati commented on component/core/src/main/java/org/exoplatform/social/core/jpa/search/ProfileIndexingServiceConnector.java:
```diff
@@ -227,6 +227,9 @@ private Document getDocument(String id) {
     fields.put("userName", identity.getRemoteId());
     fields.put("email", profile.getEmail());
     fields.put("external", String.valueOf(isExternal));
+    if (lastLoginTime != null) {
```
```suggestion
    if (profile.getProperty(Profile.LAST_LOGIN_TIME) != null) {
      fields.put("lastLoginTime", profile.getProperty(Profile.LAST_LOGIN_TIME));
    }
```

azayati commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java:
```diff
@@ -370,27 +383,43 @@ private Query getIdentitiesQuerySortedByField(String providerId,
-    if (userType != null && ( userType.equals("internal") || userType.equals("external"))) {
+    boolean isConnectedFilter = isConnected != null;
+    boolean isUserTypeFilter = userType != null && ( userType.equals("internal") || userType.equals("external"));
+    if (isConnectedFilter || isUserTypeFilter) {
```
```suggestion
    if (isConnected != null || isUserTypeFilter) {
```

azayati commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java:
```diff
@@ -370,27 +383,43 @@ private Query getIdentitiesQuerySortedByField(String providerId,
+                                                Boolean isConnected) {
     StringBuilder queryStringBuilder = null;
-    if (userType != null && ( userType.equals("internal") || userType.equals("external"))) {
+    boolean isConnectedFilter = isConnected != null;
```
not needed

azayati commented on webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementToolbar.vue:
```diff
@@ -114,9 +115,10 @@ export default {
+      this.isConnected = isConnected;
       this.userType = userType;
-      this.numberOfFilters = userType != null ? 1 : 0;
+      this.numberOfFilters = (userType != null) + (isConnected != null);
```
I prefer here that you increment the numberOfFilters variable when the condition is true instead of boolean addition

azayati replied:
@rdenarie to be confirmed for @AzmiTouil

rdenarie replied:
perhaps we should create a countFiltersMethods, and yes, I prefer a real count instead of a boolean addition

azayati reviewed (dismissed):
See review

AzmiTouil committed:
Task-42613: Advanced filters - connected / not connected users
This PR will add an advanced filter by connection status (connected or never connected) to the userManagement page.

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/identity/ProfileFilterListAccess.java:
```diff
@@ -149,6 +149,7 @@ public void setProviderId(String providerId) {
           Sorting sorting = profileFilter.getSorting();
           boolean isEnabled = profileFilter.isEnabled();
           String userType = profileFilter.getUserType();
+          Boolean isConnected = profileFilter.isConnected();
```
```suggestion
          boolean isConnected = profileFilter.isConnected();
```

AzmiTouil replied:
We need to use **Boolean** because **isConnected** can be true false or null

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java:
```diff
@@ -896,9 +896,10 @@ public int countSpaceMemberIdentitiesByProfileFilter(Space space,
                                       String sortDirection,
                                       boolean isEnabled,
                                       String userType,
+                                      Boolean isConnected,
```
```suggestion
                                      boolean isConnected,
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/IdentityDAO.java:
```diff
@@ -69,20 +69,22 @@
    * @return
    */
-  List<String> getAllIdsByProviderSorted(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean enabled, String userType, long offset, long limit);
+  List<String> getAllIdsByProviderSorted(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean enabled, String userType, Boolean isConnected, long offset, long limit);
```
```suggestion
  List<String> getAllIdsByProviderSorted(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean enabled, String userType, boolean isConnected, long offset, long limit);
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/IdentityDAO.java:
```diff
@@ -69,20 +69,22 @@
    * @return
    */
-  int getAllIdsCountByProvider(String providerId, String userType, boolean enabled);
+  int getAllIdsCountByProvider(String providerId, String userType, Boolean isConnected, boolean enabled);
```
```suggestion
  int getAllIdsCountByProvider(String providerId, String userType, boolean isConnected, boolean enabled);
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java:
```diff
@@ -142,14 +142,14 @@ public long countIdentityByProvider(String providerId) {
   @Override
-  public int getAllIdsCountByProvider(String providerId, String userType, boolean isEnabled) {
-    Query query = getIdentitiesQueryCount(providerId, userType, isEnabled);
+  public int getAllIdsCountByProvider(String providerId, String userType, Boolean isConnected, boolean isEnabled) {
```
```suggestion
  public int getAllIdsCountByProvider(String providerId, String userType, boolean isConnected, boolean isEnabled) {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java:
```diff
@@ -340,11 +340,12 @@ public int getSize() throws Exception {
   }
 
-  private Query getIdentitiesQueryCount(String providerId, String userType, boolean isEnabled) {
+  private Query getIdentitiesQueryCount(String providerId, String userType, Boolean isConnected, boolean isEnabled) {
```
```suggestion
  private Query getIdentitiesQueryCount(String providerId, String userType, boolean isConnected, boolean isEnabled) {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/IdentityDAOImpl.java:
```diff
@@ -370,27 +382,42 @@ private Query getIdentitiesQuerySortedByField(String providerId,
                                                 boolean isEnabled,
-                                                String userType) {
+                                                String userType,
+                                                Boolean isConnected) {
```
```suggestion
                                                boolean isConnected) {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/listeners/UpdateLoginTimeListenerImpl.java:
```diff
@@ -0,0 +1,38 @@
+        String username = state.getIdentity().getUserId();
+        Identity userIdentity = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, username);
+        Profile profile = userIdentity.getProfile();
+        profile.setProperty(Profile.LAST_LOGIN_TIME, Calendar.getInstance().getTime());
```
should we use the user.getLastLoginTime ?

AzmiTouil replied:
Updated!

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/listeners/UpdateLoginTimeListenerImpl.java:
```diff
@@ -0,0 +1,38 @@
+        profile.setProperty(Profile.LAST_LOGIN_TIME, Calendar.getInstance().getTime());
+        try {
+            identityManager.updateProfile(profile);
+            indexingService.reindex(ProfileIndexingServiceConnector.TYPE, userIdentity.getId());
```
I think that the updateProfile already call the reindex, so no need to reindex.
Please can you check

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/profile/ProfileFilter.java:
```diff
@@ -65,6 +65,9 @@
   private String userType;
 
+  /** Filter by user login status {connected, never connected}. */
+  private Boolean isConnected;
```
```suggestion
  private boolean isConnected;
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/profile/ProfileFilter.java:
```diff
@@ -252,6 +255,14 @@ public void setUserType(String userType) {
     this.userType = userType;
   }
 
+  public Boolean isConnected() {
```
```suggestion
  public boolean isConnected() {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/profile/ProfileFilter.java:
```diff
@@ -252,6 +255,14 @@ public void setUserType(String userType) {
+    return isConnected;
+  }
+
+  public void setConnected(Boolean isConnected) {
```
```suggestion
  public void setConnected(boolean isConnected) {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/storage/api/IdentityStorage.java:
```diff
@@ -410,11 +410,12 @@ default String getType(final String nodetype, final String property) {
    * @return
    */
-  default public List<Identity> getIdentities(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean isEnabled, String userType, long offset, long limit) {
+  default public List<Identity> getIdentities(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean isEnabled, String userType, Boolean isConnected, long offset, long limit) {
```
```suggestion
  default public List<Identity> getIdentities(String providerId, String firstCharacterFieldName, char firstCharacter, String sortField, String sortDirection, boolean isEnabled, String userType, boolean isConnected, long offset, long limit) {
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedIdentityStorage.java:
```diff
@@ -703,6 +703,7 @@ public int countSpaceMemberIdentitiesByProfileFilter(Space space, ProfileFilter
                                       String sortDirection,
                                       boolean isEnabled,
                                       String userType,
+                                      Boolean isConnected,
```
```suggestion
                                      boolean isConnected,
```

rdenarie commented on component/core/src/main/java/org/exoplatform/social/core/storage/cache/model/key/IdentityFilterKey.java:
```diff
@@ -51,6 +51,7 @@
   private String all;
   private Sorting sorting;
   private String userType;
+  private Boolean isConnected;
```
```suggestion
  private boolean isConnected;
```

rdenarie commented on component/service/src/main/java/org/exoplatform/social/rest/api/UserRestResources.java:
```diff
@@ -51,6 +52,7 @@ public abstract Response getUsers(@Context UriInfo uriInfo,
                                     @QueryParam("q") String q,
                                     @QueryParam("status") String status,
                                     @QueryParam("userType") String userType,
+                                    @QueryParam("isConnected") Boolean isConnected,
```
```suggestion
                                    @QueryParam("isConnected") boolean isConnected,
```

rdenarie commented on component/service/src/main/java/org/exoplatform/social/rest/impl/user/UserRestResourcesV1.java:
```diff
@@ -250,6 +250,7 @@ public Response getUsers(@Context UriInfo uriInfo,
                            @ApiParam(value = "User name information to filter, ex: user name, last name, first name or full name", required = false) @QueryParam("q") String q,
                            @ApiParam(value = "User status to filter online users, ex: online", required = false) @QueryParam("status") String status,
                            @ApiParam(value = "User type to filter, ex: internal, external", required = false) @DefaultValue("internal") @QueryParam("userType") String userType,
+                           @ApiParam(value = "Is connected users", required = false) @QueryParam("isConnected") Boolean isConnected,
```
```suggestion
                           @ApiParam(value = "Is connected users", required = false) @QueryParam("isConnected") boolean isConnected,
```

rdenarie commented on component/service/src/main/java/org/exoplatform/social/service/rest/PeopleRestService.java:
```diff
@@ -626,7 +626,7 @@ private void addSpaceOrUserToList(List<Identity> identities, IdentityNameList op
    * @LevelAPI Platform
    * @anchor PeopleRestService.suggestUsernames
-   * @deprecated Deprecated from 4.3.x. Replaced by a new API {@link UserRestResourcesV1#getUsers(UriInfo, String, String, String, String, boolean, int, int, boolean, String)}
+   * @deprecated Deprecated from 4.3.x. Replaced by a new API {@link UserRestResourcesV1#getUsers(UriInfo, String, String, String, Boolean, String, boolean, int, int, boolean, String)}
```
```suggestion
   * @deprecated Deprecated from 4.3.x. Replaced by a new API {@link UserRestResourcesV1#getUsers(UriInfo, String, String, String, boolean, String, boolean, int, int, boolean, String)}

```

rdenarie reviewed (changes requested):
For UpdateLoginTimeListenerImpl, did you ensure that this listener is launched AFTER the listener which set the lastLoginTime in user ?

AzmiTouil commented:
> For UpdateLoginTimeListenerImpl, did you ensure that this listener is launched AFTER the listener which set the lastLoginTime in user ?

Yes, it runs after the listener which set the lastLoginTime

rdenarie reviewed (changes requested):
Can you update the commit message 

> "Use a clear commit message of format TASK-XYZ (replace XYZ by task number) and a clear description that explains:
1/ What was the previous behavior (Prior ti this change....)
2/ What is changed/impacted by the current change
3/ (Optional) what other alternative solutions/propositions was studied"

ahmadmido committed:
Apply advanced filer only when enabled users are displayed

rdenarie committed:
Merge branch 'develop' into task-42613
